src/cargo/core/gc.rs

```diff
@@ -0,0 +1,550 @@
+///
+/// Returns None if the value is not valid. See [`parse_time_span`] if you
+/// need a variant that generates an error message.
+fn maybe_parse_time_span(span: &str) -> Option<Duration> {
+    let Some(right_i) = span.find(|c: char| !c.is_ascii_digit()) else {
+        return None;
+    };
+    let left = &span[..right_i];
+    let mut right = &span[right_i..];
+    if right.starts_with(' ') {
+        right = &right[1..];
+    }
+    let count: u64 = left.parse().ok()?;
+    let factor = match right {
+        "second" | "seconds" => 1,
+        "minute" | "minutes" => 60,
+        "hour" | "hours" => 60 * 60,
+        "day" | "days" => 24 * 60 * 60,
+        "week" | "weeks" => 7 * 24 * 60 * 60,
+        "month" | "months" => 30 * 24 * 60 * 60,
+        _ => return None,
+    };
+    Some(Duration::from_secs(factor * count))
+}
+
+/// Parses a time span string.
+pub fn parse_time_span(span: &str) -> CargoResult<Duration> {
+    maybe_parse_time_span(span).ok_or_else(|| {
+        format_err!(
+            "expected a value of the form \
+             \"N seconds/minutes/days/weeks/months\", got: {span:?}"
+        )
+    })
+}
+
+/// Parses a file size using metric or IEC units.
+pub fn parse_human_size(input: &str) -> CargoResult<u64> {
+    let re = regex::Regex::new(r"(?i)^([0-9]+(\.[0-9])?) ?(b|kb|mb|gb|kib|mib|gib)?$").unwrap();
+    let cap = re.captures(input).ok_or_else(|| {
+        format_err!(
+            "invalid size `{input}`, \
+             expected a number with an optional B, kB, MB, GB, kiB, MiB, or GiB suffix"
+        )
+    })?;
+    let factor = match cap.get(3) {
+        Some(suffix) => match suffix.as_str().to_lowercase().as_str() {
+            "b" => 1.0,
+            "kb" => 1_000.0,
+            "mb" => 1_000_000.0,
+            "gb" => 1_000_000_000.0,
+            "kib" => 1024.0,
+            "mib" => 1024.0 * 1024.0,
+            "gib" => 1024.0 * 1024.0 * 1024.0,
+            s => panic!("suffix `{s}` out of sync with regex"),
+        },
+        None => {
+            return cap[1]
+                .parse()
+                .with_context(|| format!("expected an integer size, got `{}`", &cap[1]))
+        }
+    };
+    let num = cap[1]
+        .parse::<f64>()
+        .with_context(|| format!("expected an integer or float, found `{}`", &cap[1]))?;
+    Ok((num * factor) as u64)
+}
+
+/// Performs automatic garbage collection.
+///
+/// This is called in various places in Cargo where garbage collection should
+/// be performed automatically based on the config settings. The default
+/// behavior is to only clean once a day.
+///
+/// This should only be called in code paths for commands that are already
+/// doing a lot of work. It should only be called *after* crates are
+/// downloaded so that the last-use data is updated first.
+///
+/// It should be cheap to call this multiple times (subsequent calls are
+/// ignored), but try not to abuse that.
+pub fn auto_gc(config: &Config) {
```

Comment:
This file is organized by
1. types
2. impls
3. free functions
4. tests

I found this confusing in reading it and I expect people to have false starts with this API in using it.  It appears the primary entry point into this is `gc::auto_gc`.  I saw `Gc` first and would have immediately tried to use the functions on it.   For me, I would put `gc::auto_gc` first and have it serve a a Table of Content for the rest of the file (much like the idea of putting `main` first)